packages/lesswrong/server/useractivities/getUserActivityData.ts

```diff
@@ -0,0 +1,100 @@
+import { getAnalyticsConnection } from "../analytics/postgresConnection";
+import { environmentDescriptionSetting } from "../analyticsWriter";
+
+export interface ActivityWindowData {
+  userOrClientId: string;
+  activityArray: number[];
+}
+
+/*
+ * When running this script locally we want it to use the real analytics events
+ */
+const liveEnvDescriptions: Record<string, string> = {
+  "production": 'production',
+  "staging": 'staging',
+  "dev": 'development',
+  "local-dev-prod-db": 'production', // prod running locally
+  "local-dev-staging-db": 'staging', // staging running locally
+}
+
+/**
+ * Get an array of ActivityWindowData, one for each user or client that was active between startDate and endDate.
+ */
+export async function getUserActivityData(
+  startDate: Date,
+  endDate: Date
+): Promise<ActivityWindowData[]> {
+  const analyticsDb = await getAnalyticsConnection();
+  if (!analyticsDb) {
+    throw new Error('Analytics database not available');
+  }
+  // startDate and endDate must be exact hours
+  if (startDate.getMinutes() !== 0 || startDate.getSeconds() !== 0) {
+    throw new Error('startDate must be an exact hour');
+  }
+  if (endDate.getMinutes() !== 0 || endDate.getSeconds() !== 0) {
+    throw new Error('endDate must be an exact hour');
+  }
+  if (startDate == endDate) {
+    return [];
+  }
+  if (startDate > endDate) {
+    throw new Error('startDate must be before endDate');
+  }
+  const liveEnvDescription = liveEnvDescriptions[environmentDescriptionSetting.get()]
+  if (!liveEnvDescription) {
+    throw new Error(`Unknown environmentDescriptionSetting: ${environmentDescriptionSetting.get()}`);
+  }
+  
+  const activityAnalyticsSql = `
+    WITH hourly_activity AS (
+      SELECT
+          CASE
+            WHEN event->>'userId' IS NOT NULL THEN 'u:' || (event->>'userId')
+            ELSE 'c:' || (event->>'clientId')
+          END AS "userOrClientId",
+          date_trunc('hour', timestamp) AS hour,
+          COUNT(*) AS events
+      FROM public.raw
+      WHERE environment = '${liveEnvDescription}'
+        AND event_type = 'timerEvent'
+        AND (event->>'userId' IS NOT NULL OR event->>'clientId' IS NOT NULL)
+        AND timestamp >= $1
+        AND timestamp < $2
```

Comment:
Almost all of the cost is associated with doing the sort (by time) required for SELECT DISTINCT. I've tried reasonably hard to optimise this away but without any luck, I don't want to start messing around with indexes because there isn't one that I'm confident would fix the problem and it takes such a long time to add them that trial and error would take forever (I tried adding one on the dev analytics db and gave up after 5 minutes)